### PR TITLE
Add "compact" elliptic DG operator

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This elliptic DG operator will replace the scheme that is currently shared between elliptic and evolution executables. By decoupling the elliptic from the evolution scheme the new DG operator doesn't have to treat the first-order "auxiliary" variables on the same footing as the "primal" variables anymore. Instead, it can directly apply to the primal variables now, essentially computing their second derivates, modified by the fluxes and sources of the elliptic system and by DG boundary corrections. This significantly reduces the size of the operator and therefore its memory usage and computational cost for solving it. Note that this new "compact" DG operator is actually equivalent to the first-order operator it replaces, in the sense that their solutions are exactly the same. It's the same scheme, but reduced to solving a smaller operator. This is facilitated by taking a Schur complement of the first-order operator, which is possible because the equations can be solved for the auxiliary variables just by inverting the DG mass matrix.

This PR only adds the DG operator and tests. A follow-up PR will switch the executables over to actually use the new DG operator, at which point the elliptic solver will be independent of the "boundary schemes" code that it currently shares with the evolution DG schemes.

I split the code in a few commits (essentially one commit per added file) to make it easier to review.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Comment by nilsvu** [comment]
I moved the first two commits to #2913 to make this PR slightly smaller :)

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs Since this PR is relatively involved I don't want to "request" reviews from anyone, but I really need this moving forward to get the elliptic solver merged. Would two of you be willing to take this on?

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp:128]: This pair of equations is in the opposite order from the corresponding pair above.  If there's no reason for that, swap one pair so they match.

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: Does this need a `std::move`?  I'm not sure.

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: Every `extract_subset` allocates.  I don't know if being able to do math on the resulting `Variables` outweighs that or not, but it's probably a good opportunity for another "possible optimization" comment like above.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: You can use  `<<` in ASSERT messages.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: "subitems"

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: This use was declared forbidden when `db::mutate` return values were added.  See the documentation.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Could you pass the const stuff into the mutate, like is usually done?  It doesn't look like you need the box itself in your mutation, which is usually the only case where this sort of workaround is needed.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Why explicitly list and forward some arguments, but not all of them?  (Or, can `elliptic::dg::apply_operator<System, Linearized>` be used as the argument to `db::mutate` directly?)

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:484]: I can't find this documentation.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: What is the downside to using the default?  Does it overwrite the primal fluxes mentioned in the previous paragraph?

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp:535]: Pretty much the same comments as for PrepareAndSendMortarData above.

**Review by wthrowe — CHANGES_REQUESTED** [review]
I've done this first two commits, haven't done the third yet.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:238]: Are there any checks in this block?  If not, an `INFO` doesn't do anything.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp]: The allowed errors for these tests are huge.  This margin is an order of magnitude larger than many of the values to be checked.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Review of third commit.

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: Yes I think so. If I understand correctly, copy elision only works when returning a named or temporary object:

> Standard, 12.8.32:
> - in a return statement in a function with a class return type, when the expression is the name of a non-volatile automatic object (other than a function or catch-clause parameter) with the same cv-unqualified type as the function return type, the copy/move operation can be omitted by constructing the automatic object directly into the function's return value

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: Probably not, I pushed a fixup that changes these to parameter-pack expansions and loops over components. Please let me know if you think this is a good change or if I should rather drop the fixup, I'm undecided.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: I only need the box in the mutation to retrieve arguments for the boundary conditions, in the `apply_boundary_condition` lambda. The boundary conditions are selected at runtime, so the `elliptic::apply_boundary_condition` function uses `call_with_dynamic_type` to loop over the derived classes, choose the right one and call it with arguments retrieved from the box. The only alternative to passing the box around that I can think of right now is retrieving the arguments in advance in another `call_with_dynamic_type`, storing const-refs in a large `std::variant` (one variant for every registered boundary-condition class) and passing that to `prepare_mortar_data`.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Pushed a fixup that forwards all const-ref args. Is there a way to get rid of the lambda entirely, or forward all arguments including the not-null pointers? Just passing `&elliptic::dg::apply_operator<System, Linearized>` instead of the lambda doesn't work because the function infers its template parameters from arguments.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: The linear and nonlinear operator apply to different variables, so when using the defaults here they each set up memory buffers for their mortar data. However, when the operator applications don't overlap in the algorithm, as is the case for the Newton-Krylov algorithm, we can just re-use the same buffers in the DataBox for both the linear and nonlinear operator. In that case we can set these template parameters to the same tags. I extended the docs here a little bit.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:238]: The `INFO` also prints when Catch2 fails a test for other reasons, such as uncaught exceptions. I think I put this in because I had a bug here at some point, but I don't remember exactly. I can take it out if you prefer.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp]: I'll do a few more tests to look at the error margins here.

**Comment by nilsvu** [comment]
I appreciate you finding the time to review this @wthrowe! I pushed fixups to address most of your comments and commented on the others.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Sometimes you can get rid of the lambda with a cast to a function pointer type, but I don't see a way to work out what the correct type to cast to is here.  You can forward the pointers by just removing the explicit arguments and absorbing them into the argument pack.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: OK, yep, I missed the explicit use of the box.  Fine as-is.

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (wthrowe) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:238]: OK, that's fine.

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Maybe move the objects out of the box, mutate them, and then move them back in?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Added a fixup that does that. Do you think that's preferable to retrieving the pointers?

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: Of course 🤦‍♂️ the pointers can also just be forwarded as const-refs..

**Comment by nilsvu** [comment]
Pushed two more fixup. Thanks @wthrowe!

File: src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp
Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/Actions/ApplyOperator.hpp]: In a case where there might be compute items involved, I think it is better to always keep the box in a consistent state by moving the data out.  For the specific case where there are known to be no compute items I'm less sure, but there was a consensus among the reviewers on #2933 that pulling out pointers was bad, so I'll follow that until there's a reason not to.

**Comment by wthrowe** [comment]
I'm happy with the current state except for the open question about test tolerances.  Go ahead and squash the rest.

**Comment by nilsvu** [comment]
Rebased and squashed the fixups. I pushed another fixup that addresses the issue with test tolerances. It adjusts the scales and adds additional higher-resolution analytic-solution tests with stricter tolerances. Please take a look when you get the chance :)

File: tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: I'm still worried that you need to accept 80% errors in this case (and slightly less above).  Why is this result so inaccurate?

Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Well this 3D case tries to resolve a full sinusoid period with 2 linear elements in xi-direction :) At x=-0.5 the gradient is zero analytically, but 2 numerically because the sinusoid is approximated as a ^ shape (or upside-down V, i.e. just two lines with slope 2 and -2). That's a large error. I chose so few elements to avoid writing even more numbers in the regression test. The added code below repeats the analytic-solution test with much higher resolution. Note that the low tolerance here only applies to the analytic-solution test; the custom values are checked to default precision (1.-14 or so).

Comment (wthrowe) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Can you set the expected values to what you actually expect the operator to return, then?

Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Yea that's what the regression test does on the low-resolution domain. I could disable the analytic-solution test on the low-resolution domain by making the `approx` arguments `std::optional`s, so the comparison to the analytic solution only runs on the high-resolution domain. Not sure if that's worth it though, it just deletes these low-resolution/high-tolerance analytic-solution tests.

Comment (wthrowe) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Sorry, you've lost me.  This *is* the regression test, isn't it?

Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Yes it is. It compares the hard-coded numbers to operator output at default precision (1e-14 or so). It is at low resolution to keep the amount of hard-coded numbers low, and because the resolution is irrelevant for the regression test. It also does an analytic-solution test because that's built-in to the testing function, but does it at very low resolution and thus very high tolerance. Therefore, as response to your comment, I added another test at much higher resolution below. That one doesn't test any hard-coded numbers but only the analytic solution at much smaller tolerance.

Comment (wthrowe) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Oh, I see now.  Could you add a comment explaining that these loose approx values are effectively disabling a part of the test you don't want, but there is another part running with more typical tolerances?  I didn't understand that and thought these were the only tolerances being used.

Comment (nilsvu) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: Done! Pushed another fixup.

Comment (wthrowe) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/Test_DgOperator.cpp:711]: I think that's clear, now.  Go ahead and squash.

**Comment by nilsvu** [comment]
Rebased on #3048.

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsvu** [comment]
Rebased again :)

**Comment by nilsvu** [comment]
@kidder could you perhaps do the second review for this or help find a second reviewer?

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (nilsdeppe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: `every` -> `ever`

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp
Comment (nilsdeppe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: the documentation said you use `N_points`, is `p=N_points` here? That would be good to clarify :)

**Review by nilsdeppe — DISMISSED** [review]
LGTM, great documentation! A couple minor comments, feel free to ignore :)

**Comment by nilsvu** [comment]
Rebased and fixed those two things. This is a fairly involved PR, so your reviews are much appreciated @wthrowe and @nilsdeppe :)

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
